### DISCUSSION TITLE
Callbacks from UI

### DISCUSSION THREAD

**Discussion by ejs-ejs** [IfcOpenShell/IfcOpenShell#4191] [Q&A] [answered]
Hi All,
Sorry for the strange question, but I have a problem I can't find a solution. Long story short: I need to rename parameters for specific IFC element types, like IfcCovering and IfcFlowSegment. The source data comes from MagiCAD@AutoCAD, and we have no competence and time left to export it properly.

So, I'm expanding the 'bulk property editing' feature to allow it to alter the selected elements only. An extra setting is added to the UI, but I can't find a callback from the UI to be patched. The debug from ```bim.add_property_to_edit``` shows that ```pset_name``` , ```existing_property_name``` and ```new_property_name``` are assigned, but ```selected_only``` is not. The actual renaming of the selected parameters seems to work flawlessly.

The rough code of what I should achieve is:
```
def _execute(self, context):
        props_to_map = context.scene.RenameProperties
        
        if props_to_map.selected_only:
            ifc_elements = list()
            for obj in tool.Blender.get_selected_objects():
                ifc_definition_id = obj.BIMObjectProperties.ifc_definition_id
                if not ifc_definition_id:
                    continue
                ifc_elements.append(tool.Ifc.get().by_id(ifc_definition_id))
        else:
            ifc_file = IfcStore.get_file()
            ifc_elements = ifc_file.by_type("IfcElement")

        for ifc_element in ifc_elements:
            for definition in ifc_element.IsDefinedBy:
                if definition.is_a("IfcRelDefinesByProperties"):
                    prop_set = definition.RelatingPropertyDefinition
                    self.rename_property(prop_set, props_to_map, ifc_element)
        self.report({"INFO"}, "Finished applying changes")
        return {"FINISHED"}
```

The classes are:
```
class RenameProperties(PropertyGroup):
    pset_name: StringProperty(name="Pset")
    existing_property_name: StringProperty(name="Existing Property Name")
    new_property_name: StringProperty(name="New Property Name")
    selected_only: BoolProperty(name="For selected only", description="Apply only for the selected entities", default = True)
```
and 

```
class BIM_PT_rename_parameters(Panel):
   ...

    def draw(self, context):
        layout = self.layout
        props = context.scene.RenameProperties
        row = layout.row()
        op = row.operator("bim.add_property_to_edit", icon="ADD")
        op.option = "RenameProperties"
        if props:
           ...
            row = layout.row(align=True)
            row.prop(prop, "selected_only")
            row.operator("bim.rename_parameters", icon="CHECKMARK")
            clear = row.operator("bim.clear_list", icon="CANCEL", text="")
            clear.option = "RenameProperties"
```

Sorry for the messy text; I will try to improve it.

**Comment by Moult** [comment]
The code looks correct. I think I'm misunderstanding what the problem is. What exactly is the error?

**Comment by ejs-ejs** [comment, marked as answer]
Sorry for the wrong post. Seems like everything is working fine today. 
The error was in ```operator.py``` like , "props_to_map.selected_only" is not available. 

Maybe it was just one of these Late Night Problems.
